### ISSUE TITLE
Support for custom ranks in Wikidata schema

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#2326]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**

Wikidata statements can have three ["ranks"](https://www.wikidata.org/wiki/Help:Ranking): Preferred, Normal or Deprecated.

When creating Wikidata statements from OpenRefine, in the Wikidata schema, it is not currently possible to set a custom rank for the statements created: the rank is always "normal".

There is already an image for ranks in the UI but is not clickable.

**Describe the solution you'd like**

It would be useful if the rank image in each statement could be clickable and one could define a custom rank, similarly to the way it is done on Wikidata itself.

**Additional context**

The rank icon looks like this:  ![image](https://user-images.githubusercontent.com/309908/75116180-8e71ee80-565d-11ea-9bc4-b1040038c812.png). This represents the three ranks (the grey filling moves depending on which rank is active).

Here is what it looks like in Wikidata:
![image](https://user-images.githubusercontent.com/309908/75116149-418e1800-565d-11ea-9989-87bb32191878.png)
When editing the statement:
![image](https://user-images.githubusercontent.com/309908/75116156-51a5f780-565d-11ea-82e0-2c1b3d254743.png)

Current UI in OpenRefine:
![image](https://user-images.githubusercontent.com/309908/75116167-74381080-565d-11ea-8c89-dc0cd169dd48.png)

**Comment by thadguidry** [comment]
Can you further elaborate on the issue what exactly happens in the UI when a user would click on the rank icon in OpenRefine ?  Some popup or dropdown or what , since you mention "custom" is it some input field that opens?

**Comment by wetneb** [comment]
I would say just mimic what Wikidata's own UI does - that is the most precise description I can give :)
You can just try it out by yourself at https://www.wikidata.org/wiki/Q4115189: edit any statement and change the rank there.

**Comment by thadguidry** [comment]
Gotcha, understood now.

**Comment by Bhanu-mbvg** [comment]
Hey, can I work on this issue?

**Comment by thadguidry** [comment]
@Bhanu-mbvg sure!

**Comment by Bhanu-mbvg** [comment]
Hey, can I get to know how I can recreate this issue on my OpenRefine server?

**Comment by Bhanu-mbvg** [comment]
Hey, can you guide me on where can I get the code for the Schema page?

**Comment by VojtechDostal** [comment]
@Bhanu-mbvg Lovely! This improvement would really make OpenRefine the most powerful mass-editing tool for Wikidata - Neither QuickStatements nor Petscan can change ranks AFAIK.

**Comment by jinoytommanjaly** [comment]
Any updates on this. Anyone working on this to add custom ranking.?

**Comment by SIXIN-LI** [comment]
Hi, can I work on this?

For the solution, do you want exactly the same three ranks as Wikidata or additional ranks that users can define? I feel like the three ranks provided are good enough, thinking that the rank is only time-based.

For now, Wikidata is not ordering multiple values based on their ranks while they use color to identify them. This may be something we can add.

What do you think?

**Comment by wetneb** [comment]
> For the solution, do you want exactly the same three ranks as Wikidata or additional ranks that users can define? I feel like the three ranks provided are good enough, thinking that the rank is only time-based.

Users will only be able to upload statements with ranks that are supported by Wikibase, so I would not propose any other option, otherwise people will run into errors at upload time. I do not think ranks are extensible: all Wikibase instances support just these three ranks, I think (unless I missed something).

> For now, Wikidata is not ordering multiple values based on their ranks while they use color to identify them. This may be something we can add.

I think using color to identify them is indeed quite useful, otherwise it is hard to notice the difference. Note that we are already using colors to distinguish between added and removed statements.

**Comment by SIXIN-LI** [comment]
Hi @wetneb, 

I believe I need to add a property named "rank" to the Statement class. I have located the class StatementEdit and the Statement class seems in a jar file "wdtk-datamodel-0.13.1". I wonder how I can modify the library.

**Comment by wetneb** [comment]
I think you should not need to modify the library to work on this. WDTK (Wikidata-Toolkit) indeed already supports statement ranks, this task is about exposing this support to OpenRefine users.

**Comment by SIXIN-LI** [comment]
Thank you for the clarification! 

For the logic of toggling the icon style, I wonder if we can do it by clicking the up arrow/center circle/down arrow. This will save one click and is mostly used but will need to assume users are familiar with these ranks in Wikidata for them.

The way Wikidata does now is that after clicking, it gives a drop-down menu, and via options in the menu, it toggles the icon style.

**Comment by SIXIN-LI** [comment]
@wetneb 
My idea is something like the following gif. When the user hovers over a certain part, the title showing the rank value will show. 

 Does it seem fine to you?
![rankToggle](https://user-images.githubusercontent.com/45926850/161511471-6f5542c5-dca2-4715-8918-be8d2fb5eeaf.gif)

**Comment by SIXIN-LI** [comment]
Thanks! I had a hard time interpreting the backend structure. I wonder what the workflow in the backend is like after the item JSON value is passed to the backend.

**Comment by SIXIN-LI** [comment]
Hi @wetneb , from a wikibase doc in the project and the codes in the wikidata folder, I guess the pipeline to add rank to the statement is as below. Do you mind checking if my understanding is correct?

Pipeline:
1. Create a WbRankConstant class that extends WbExpression class
2. Make WbRankConstant a parameter of the WbStatementExpr's constructor
3. May need to scrutinize rank? (I am not sure if this is needed)
4. To prepare for identifying updates, create an entity value class for rank, and add rank to PointerExtractor class
5. I didn't locate the classes responsible to compare edits with existing data. I am not sure if the scheduler is doing this.
6. Merging step: create a rankMerger class

**Comment by antoine2711** [comment]
Hi @SIXIN-LI : I can't answer all your questions, but I'm pretty sure that OR doesn't read Wikidata before pushing. I think it pushes, and Wikidata does nothing if there no change to data in Wikidata.
See « #4624 Adding a statement that already exists on Wikibase should not result in any API action ».

> 5. I didn't locate the classes responsible to compare edits with existing data. I am not sure if the scheduler is doing this.

Regards, Antoine

**Comment by SIXIN-LI** [comment]
> Hi @SIXIN-LI : I can't answer all your questions, but I'm pretty sure that OR doesn't read Wikidata before pushing. I think it pushes, and Wikidata does nothing if there no change to data in Wikidata.
> 
> See « #4624 Adding a statement that already exists on Wikibase should not result in any API action ».
> 
> > 5. I didn't locate the classes responsible to compare edits with existing data. I am not sure if the scheduler is doing this.
> 
> Regards, Antoine

Thank you for the clarification! I am getting there, except needing a CSV file to do the testing.

**Comment by antoine2711** [comment]
> Thank you for the clarification! I am getting there, except needing a CSV file to do the testing.

I suggest you use data from this folder if you can. It's the data for the « Functional tests », the user interface test cases & units…
`/main/tests/data`

Regards, Antoine

**Comment by wetneb** [comment]
@SIXIN-LI I think you have a very good understanding of the structure!

> May need to scrutinize rank? (I am not sure if this is needed)

That is indeed optional, and only if we identify meaningful problems involving ranks that should be surfaced.

> To prepare for identifying updates, create an entity value class for rank, and add rank to PointerExtractor class

I suspect you might just be able to use Wikidata-Toolkit's existing `enum` to represent the rank. Since ranks do not refer to any particular item, they should be ignored by `PointerExtractor`.

> I didn't locate the classes responsible to compare edits with existing data. I am not sure if the scheduler is doing this.

That is happening in the `org.openrefine.wikidata.schema.strategies` namespace.

**Comment by SIXIN-LI** [comment]
> I think you have a very good understanding of the structure!

Thank you for the affirmative answer! 

> That is happening in the org.openrefine.wikidata.schema.strategies namespace.

I found all three strategies are using the existing rank when merging. If there is a conflict between the rank of the existing statement and that of the to-add statement, should the new rank be taken instead?

**Comment by wetneb** [comment]
> If there is a conflict between the rank of the existing statement and that of the to-add statement, should the new rank be taken instead?

That as an extremely interesting question! For this, I think it is important to think about a couple of concrete data import scenarios, think about the desired behaviour in those scenarios, and then figure out a way to make those fit in OpenRefine. It is imaginable to add options to the merging strategies if needed.

**Comment by SIXIN-LI** [comment]
> But I'm pretty sure that OR doesn't read Wikidata before pushing. I think it pushes, and Wikidata does nothing if there no change to data in Wikidata.

Hi @antoine2711 , I found that in EditBatchProcessor class which prepares for edits uses the WikibaseDataFetcher to fetch the current documents for this batch of updates from Wikibase via the online API. So can I understand it as OR is fetching data from wikibase before performing edits? Or is there another way to understand the fecther?

**Comment by antoine2711** [comment]
> Hi @antoine2711 , I found that the EditBatchProcessor class which prepares for edits uses the WikibaseDataFetcher to fetch the current documents for this batch of updates from Wikibase via the online API. So can I understand it as OR is fetching data from wikibase before performing edits? Or is there another way to understand the fecther?

Let's ask @wetneb as I'm not very good on the Java side, I can't answers you with certitude.

Regards, Antoine

**Comment by wetneb** [comment]
> Hi @antoine2711 , I found that the EditBatchProcessor class which prepares for edits uses the WikibaseDataFetcher to fetch the current documents for this batch of updates from Wikibase via the online API. So can I understand it as OR is fetching data from wikibase before performing edits? Or is there another way to understand the fecther?

That is correct: before performing edits on existing entities, we fetch the current contents of those entities, to decide how to integrate the edit into the entity (according to the specified merging strategy).

**Comment by SIXIN-LI** [comment]
> > If there is a conflict between the rank of the existing statement and that of the to-add statement, should the new rank be taken instead?
> 
> That as an extremely interesting question! For this, I think it is important to think about a couple of concrete data import scenarios, think about the desired behaviour in those scenarios, and then figure out a way to make those fit in OpenRefine. It is imaginable to add options to the merging strategies if needed.

Hi @wetneb, I can think that most of the time users change the ranks of statistics like population or life expectancy, or recurring events. If they revise the rank in batch, then the to-add rank should override the existing rank since users mean to update. Do you see more use cases to consider? 

For the default merger and snakOnly merger, they both cover changes in the main value. So I think it is reasonable to make the merged item with a to-add rank. 

Also, I am a bit confusing since I didn't find a place on the wikibase schema revising page or uploading edit page for OR users to change the merger and editing mode. I do see the default setting is default merger and add_or_merge mode. Have merging strategies not been formally introduced to the public yet?

**Comment by wetneb** [comment]
> Hi @wetneb, I can think that most of the time users change the ranks of statistics like population or life expectancy, or recurring events. If they revise the rank in batch, then the to-add rank should override the existing rank since users mean to update. Do you see more use cases to consider?

Say I am importing population counts from a new version of the census for some country. I will add statements with "point in time" qualifiers, and I will probably want that my statements have a "preferred" rank, because they are the newest ones. At the same time, I probably want to put any other statement that had preferred rank back to "normal" rank (because they are not the latest anymore).

One way to achieve this would be to use data extension to fetch the current preferred values, and then add them back in the schema with normal rank. But that would also require fetching the qualifiers on the corresponding statements (which is not currently possible with the Wikibase reconciliation service), to make sure we are able to identify the statements correctly when resetting their rank.

As a user I would find it much easier if that could be done "automatically" in some way. But perhaps the general rule of "downgrade any existing statement with preferred rank to normal rank when adding a new preferred rank statement" is not helpful in other cases? When does it make sense to have multiple statements with preferred rank?
Also, this reasoning is based on my experience in Wikidata, but perhaps other Wikibase instances use ranks in different ways?

> Also, I am a bit confusing since I didn't find a place on the wikibase schema revising page or uploading edit page for OR users to change the merger and editing mode. I do see the default setting is default merger and add_or_merge mode. Have merging strategies not been formally introduced to the public yet?

This is only available in the master branch, not in the latest stable (3.5.2) release, but yes, in master it is exposed to the user. You can click the "configure" link in any statement in the schema to access those options.

**Comment by SIXIN-LI** [comment]
@wetneb Fetching and modifying the statements that had preferred ranks may not be helpful when there can be multiple preferred statements. One example is diverging information provided by different sources. In this case, one source may not downgrade info uploaded by another source.

So, when offering the ranking modifying feature, we may consider adding an option for users to decide if they want to override the existent statements with preferred ranks or not. This should give OR some flexibility to apply to use cases that we haven't thought of.

> perhaps other Wikibase instances use ranks in different ways?

I checked several notable wikibase instances shown here https://wikiba.se/showcase/. FactGrid and Europena Eagle is using the rank similar to Wikidata while most of the others don't list deprecated info and latest info together, from which I assumed they do not need to use the rank attribute to differentiate Infos. Rhizome Artbase does have time-aligned descriptions for one art piece. Without looking into its code, it is hard to tell if they add the description by creation time or using rank. Are there other major wikibase instances that I should have a look at?

**Comment by wetneb** [comment]
Wow, thank you for the in depth review of the question! I think you have a very good understanding of the issue and it is not worth looking at more Wikibase instances.
I agree about the case where there are multiple preferred statements.

**Comment by SIXIN-LI** [comment]
> It is not worth looking at more Wikibase instances

Yeah, it seems like Wikidata is the most notable Wikibase instance.

@wetneb Having another thought about uploading updated ranks in certain cases(described below), if uploading customized ranks brings more inefficiency than valid assistance to the whole editing process, I am not sure if we still need it.

Say, if users only update a statement once a year(or a long time) and if they do not keep track of their contributions, the next period when they try to update it again and they are not sure if the preferred data was uploaded by themselves, by playing safe they would not mark the existent data with preferred ranks as normal, and then upload their data with preferred ranks. 

Another case is when some preferred statements are uploaded by themselves and some are not, (I assume) users would not check every property that they need to update so they would play safe to not update existing ranks and then upload their data with preferred ranks. With such acts happening over and over again, statements with preferred ranks will stack and this makes preferred rank less helpful.
And this kind of inefficiency would not happen if they do not update the ranks the first time.

I am new to wikidata. So I cannot speak on the behalf of most users. But if wikidata users are all discrete about the data they uploaded, they may prefer safe cards from the beginning.

If the cases that I mention are only a trivial part, then we can continue with offering options for users to decide whether they want to modify existing statements' ranks.

**Comment by SIXIN-LI** [comment]
Hi @wetneb, is my consideration above reasonable? How should I proceed?

Moreover, the front-end logic of showing options for users to decide whether they want to modify existing statements' ranks may need some more insights.

**Comment by wetneb** [comment]
> Say, if users only update a statement once a year(or a long time) and if they do not keep track of their contributions, the next period when they try to update it again and they are not sure if the preferred data was uploaded by themselves, by playing safe they would not mark the existent data with preferred ranks as normal, and then upload their data with preferred ranks.

I am not sure why it is important that they know if they are the author of the current preferred statement or not. Statement ranks are normally set following a set of rules, not by taste, so the authorship should not really matter, should it?

> Moreover, the front-end logic of showing options for users to decide whether they want to modify existing statements' ranks may need some more insights.

Yes I feel like this requires more design indeed!
If we want to "simply" add support for ranks for new statements, then we should be clear about what happens to those ranks when those statements are merged with existing ones.
However I am still worried that we are just "adding support" for the sake of it without really considering users' needs here.

I feel bad about having labelled this with `good second issue` since the design effort involved is actually quite big.

**Comment by wetneb** [comment]
Here is one other use case worth considering when designing how the edited ranks should be merged with existing ranks.

When doing a data import of some identifiers, I checked existing identifier statements on Wikidata and noticed that some of them were invalid (because they did not exist in the third-party database anymore). So I wanted to mark those statements with a deprecated rank, and add a qualifier to indicate the reason for deprecation.

This means transforming statements from this:
![image](https://github.com/OpenRefine/OpenRefine/assets/309908/93e1f626-cdfe-49ca-8897-26dc4dc1664d)
to that:
![image](https://github.com/OpenRefine/OpenRefine/assets/309908/015333f4-c560-47a1-8ce7-5f20546a835e)
See [a sample edit](https://www.wikidata.org/w/index.php?title=Q153006&diff=1911349355&oldid=1902910428) doing that (in two steps, since OpenRefine does not support doing that yet).

As a user I would simply expect to set the rank in the schema, add the qualifier in the same statement, and upload the changes to perform both actions (rank change and qualifier addition) in one go.
